main_OSEM.py

```diff
@@ -0,0 +1,102 @@
+            raise StopIteration
+
+
+class Submission(Algorithm):
+    """
+    This class provides an implementation of the OSEM algorithm (marginally modified).
+    Of course, OSEM does not converge to the MAP solution for this Challenge
+    (you will have to use data.prior for that), but hopefully this shows you how to
+    implement an algorithm from baree-bones.
+    In this implementation, we have chosen NOT to use the sirf.STIR Poisson objective
+    function and its functionality. This is illustrated in the BSREM examples on
+    https://github.com/SyneRBI/SIRF-Contribs/tree/master/src/Python/sirf/contrib/BSREM
+    
+    Note that in this implentation, we use a particular feature of OSEM (and in fact,
+    of the Poisson log-likelihood), that the multiplicative term cancels in the quotient
+    of measured and estimated data (thanks to our convention for the additive_term).
+    It turns out you only need it for the "sensitivity" terms.
+    """
+    def __init__(self, data: Dataset, num_subsets: int = 7, update_objective_interval: int = 10, **kwargs):
+        """
+        Initialisation function, setting up data & (hyper)parameters.
+        NB: in practice, `num_subsets` should likely be determined from the data.
+        This is just an example. Try to modify and improve it!
+        """
+        self.acquisition_models = []
+        self.prompts_subsets = []
+        self.subset_sensitivities = []
+        self.subset = 0
+        self.x = data.OSEM_image.clone()
+    
+        # find views in each subset
+        # (note that SIRF can currently only do subsets over views)
+        views = data.mult_factors.dimensions()[2]
+        partitions_idxs = partition_indices(num_subsets, list(range(views)), stagger=True)
+
+        # for each subset: find data, create acq_model, and create subset_sensitivity (backproj of 1)
+        for i in range(num_subsets):
+            prompts_subset = data.acquired_data.get_subset(partitions_idxs[i])
+            additive_term_subset = data.additive_term.get_subset(partitions_idxs[i])
+            multiplicative_factors_subset = data.mult_factors.get_subset(partitions_idxs[i])
+        
+            acquisition_model_subset = STIR.AcquisitionModelUsingParallelproj()
+            acquisition_model_subset.set_additive_term(additive_term_subset)
+            acquisition_model_subset.set_up(prompts_subset, self.x)
+
+            subset_sensitivity = acquisition_model_subset.backward(multiplicative_factors_subset)
+            # add a small number to avoid NaN in division
+            subset_sensitivity += subset_sensitivity.max() * 1e-6
+
+            self.acquisition_models.append(acquisition_model_subset)
+            self.prompts_subsets.append(prompts_subset)
+            self.subset_sensitivities.append(subset_sensitivity)
+
+        super().__init__(update_objective_interval=update_objective_interval, **kwargs)
+        self.configured = True # required by Algorithm
+
+    def update(self):
```

Comment:
this currently only works on one subset only, so is a "subiteration" rather than "iteration", right?

I suppose that's ok if we are fine with triggering a timeout mid-subset.

Reply:
This is true for any of the subset algorithms. It's why I don't like the CIL terminology of "iteration" and prefer "update", but clearly impossible to change that.

Reply:
> I suppose that's ok if we are fine with triggering a timeout mid-subset.

I guess you mean "mid-full iteration". Yes, we don't know what the algorithm is doing, so we can only act whenever `next` is called.

Reply:
I mean there are 2 options:

1. (current) `update()` does one subset, then `subset += 1`
2. (alternative) `update()` does `for i in range(num_subsets)`

and IMO the sole reason to chose (1) is it allows timeout to trigger "mid-full iteration". But perhaps we should forbid that and actually do (2) instead?

Reply:
I was asking about the `why` not the `how`. Thx for the answer.